e2e/connect/acls.go

```diff
@@ -0,0 +1,401 @@
+	fakeToken := uuid.Generate()
+	job := tc.parseJobSpecFile(t, demoConnectJob)
+
+	jobAPI := tc.Nomad().Jobs()
+
+	// deliberately set the fake Consul token
+	job.ConsulToken = &fakeToken
+
+	// should fail, because the token is fake
+	_, _, err := jobAPI.Register(job, nil)
+	r.Error(err)
+	t.Log("job correctly rejected, with error:", err)
+}
+
+func (tc *ConnectACLsE2ETest) TestConnectACLs_ConnectDemo(f *framework.F) {
+	t := f.T()
+	r := require.New(t)
+
+	t.Log("test register Connect job w/ ACLs enabled w/ operator token")
+
+	// === Setup ACL policy and token ===
+
+	// create a policy allowing writes of services "count-api" and "count-dashboard"
+	policyID := tc.createConsulPolicy(consulPolicy{
+		Name:  "nomad-operator-policy",
+		Rules: `service "count-api" { policy = "write" } service "count-dashboard" { policy = "write" }`,
+	}, f)
+	t.Log("created operator policy:", policyID)
+
+	// create a Consul "operator token" blessed with the above policy
+	operatorToken := tc.createOperatorToken(policyID, f)
+	t.Log("created operator token:", operatorToken)
+
+	// === Register the Nomad job ===
+
+	// parse the example connect jobspec file
+	jobID := "connect" + uuid.Generate()[0:8]
+	tc.jobIDs = append(tc.jobIDs, jobID)
+	job := tc.parseJobSpecFile(t, demoConnectJob)
+	job.ID = &jobID
+	jobAPI := tc.Nomad().Jobs()
+
+	// set the valid consul operator token
+	job.ConsulToken = &operatorToken
+
+	// registering the job should succeed
+	resp, _, err := jobAPI.Register(job, nil)
+	r.NoError(err)
+	r.NotNil(resp)
+	r.Empty(resp.Warnings)
+	t.Log("job has been registered with evalID:", resp.EvalID)
+
+	// === Make sure the evaluation actually succeeds ===
+EVAL:
+	qOpts := &napi.QueryOptions{WaitIndex: resp.EvalCreateIndex}
+	evalAPI := tc.Nomad().Evaluations()
+	eval, qMeta, err := evalAPI.Info(resp.EvalID, qOpts)
+	r.NoError(err)
+	qOpts.WaitIndex = qMeta.LastIndex
+
+	switch eval.Status {
+	case "pending":
+		goto EVAL
+	case "complete":
+	// ok!
+	case "failed", "canceled", "blocked":
+		r.Failf("eval %s\n%s\n", eval.Status, pretty.Sprint(eval))
+	default:
+		r.Failf("unknown eval status: %s\n%s\n", eval.Status, pretty.Sprint(eval))
+	}
+
+	// assert there were no placement failures
+	r.Zero(eval.FailedTGAllocs, pretty.Sprint(eval.FailedTGAllocs))
+	r.Len(eval.QueuedAllocations, 2, pretty.Sprint(eval.QueuedAllocations))
+
+	// === Assert allocs are running ===
+	for i := 0; i < 20; i++ {
+		allocs, qMeta, err := evalAPI.Allocations(eval.ID, qOpts)
+		r.NoError(err)
+		r.Len(allocs, 2)
+		qOpts.WaitIndex = qMeta.LastIndex
+
+		running := 0
+		for _, alloc := range allocs {
+			switch alloc.ClientStatus {
+			case "running":
+				running++
+			case "pending":
+				// keep trying
+			default:
+				r.Failf("alloc failed", "alloc: %s", pretty.Sprint(alloc))
+			}
+		}
+
+		if running == len(allocs) {
+			break
+		}
+
+		time.Sleep(500 * time.Millisecond)
+	}
+
+	allocs, _, err := evalAPI.Allocations(eval.ID, qOpts)
```

Comment:
Don't we already have these allocs from the previous loop?

Reply:
:+1: